app/com/arpnetworking/kairos/service/KairosDbServiceImpl.java

```diff
@@ -0,0 +1,256 @@
+        } else {
+            containsFilter = s -> true;
+        }
+
+        return metricNames.stream()
+                .filter(baseFilter)
+                .filter(IS_PT1M.negate())
+                .filter(containsFilter)
+                .collect(Collectors.collectingAndThen(Collectors.toList(), ImmutableList::copyOf));
+    }
+
+
+    private CompletionStage<List<String>> getMetricNames() {
+        final List<String> metrics = _cache.getIfPresent(METRICS_KEY);
+
+        final CompletionStage<List<String>> response;
+        if (metrics != null) {
+            response = CompletableFuture.completedFuture(metrics);
+        } else {
+            // TODO(brandon): Investigate refreshing eagerly or in the background
+            // Refresh the cache
+            final CompletionStage<List<String>> queryResponse = _kairosDbClient.queryMetricNames()
+                    .thenApply(KairosMetricNamesQueryResponse::getResults)
+                    .thenApply(list -> {
+                        _cache.put(METRICS_KEY, list);
+                        _metricsList.set(list);
+                        return list;
+                    });
+
+            final List<String> metricsList = _metricsList.get();
+            if (metricsList != null) {
+                response = CompletableFuture.completedFuture(metricsList);
+            } else {
+                response = queryResponse;
+            }
+        }
+
+        return response;
+    }
+
+    private static MetricsQuery useAvailableRollups(final List<String> metricNames, final MetricsQuery originalQuery) {
+        final MetricsQuery.Builder newQueryBuilder = new MetricsQuery.Builder()
+                .setStartTime(originalQuery.getStartTime());
+
+        originalQuery.getEndTime().ifPresent(newQueryBuilder::setEndTime);
+
+
+        newQueryBuilder.setMetrics(originalQuery.getMetrics().stream().map(metric -> {
+            // Check to see if there are any rollups for this metrics
+            final String metricName = metric.getName();
+            final ImmutableList<String> filteredMetrics = filterMetricNames(metricNames, Optional.of(metricName), false);
+            final List<String> rollupMetrics = filteredMetrics
+                    .stream()
+                    .filter(IS_ROLLUP)
+                    .filter(s -> s.length() == metricName.length() + 3)
+                    .collect(Collectors.toList());
+
+            if (rollupMetrics.isEmpty()) {
+                // No rollups so execute what we received
+                return metric;
+            } else {
+                // There are rollups, now determine the appropriate one based on the max sampling period in the
+                // aggregators
+                final Optional<SamplingUnit> maxUnit = metric.getAggregators().stream()
+                        .filter(agg -> agg.getAlignSampling().orElse(Boolean.FALSE)) // Filter out non-sampling aligned
```

Comment:
It would be nice to understand the breakdown of queries not leveraging roll-ups and why (e.g. via metrics).

Reply:
Done